File: genetic-packer/src/main/java/genetic/packer/evolution/fitness/impl/BaseFitnessTranslator.java
package genetic.packer.evolution.fitness.impl;

import javaslang.collection.Traversable;

/**
 * @author piotr.larysz
 */
public abstract class BaseFitnessTranslator {

    protected double rateIntersections(Traversable<Boolean> intersections) {
        return intersections
                .map(this::rateSingleIntersection)
                .reduce((left, right) -> left * right);
    }

    private double rateSingleIntersection(boolean intersects) {
        return intersects ? 1 : 1.5;
    }
}

File: genetic-packer/src/main/java/genetic/packer/evolution/fitness/impl/BaseFitnessTranslator.kt
package genetic.packer.evolution.fitness.impl

abstract class BaseFitnessTranslator {

  protected fun rateIntersections(intersections: Sequence<Boolean>) = intersections
      .map { rateSingleIntersection(it) }
      .reduce { left, right -> left * right }

  private fun rateSingleIntersection(intersects: Boolean) = if (intersects) 1.0 else 1.5
}